### ISSUE TITLE
Add conv_layer_spatial.cl

### ISSUE THREAD

**Issue by edgarriba** [edgarriba/tiny-cnn#2]
I'm trying to implement spatial convolution based on Caffe kernels.
We assume as a first approach to have only defined `MULTI`

The implementation can be found here
https://github.com/edgarriba/tiny-cnn/blob/opencl-ops-patch/tiny_cnn/core/kernels/conv2d_opencl_op.h#L57

and kernels definition
https://github.com/edgarriba/tiny-cnn/blob/opencl-ops-patch/tiny_cnn/layers/convolutional_layer.h#L301

After launching the kernel we don't get errors but the output values are not the expected ones.
We suspect that compile definitions are not properly set or that we need additional preprocessing in weights as is done [here](https://github.com/naibaf7/caffe/blob/master/src/caffe/layers/conv_layer_spatial.cpp#L585)

@naibaf7 @gongzg
Did you face similar problems in Caffe integration?

**Comment by naibaf7** [comment]
@edgarriba 
Yes I believe you can't properly use the spatial convolution kernel without the whole tuning & preprocessing logic that goes on in the Caffe integration.

I am planning to integrate the logic/idea of the intel spatial kernel (which is to rely on Intel's instructions to exchange data between subgroups rather than doing a round trip to the system memory each time) smoothly into LibDNN, but as my Skylake GT platform is still not working properly with Intel's spatial kernels, this project is shifted back until driver (beignet) and LLVM/CLANG issues are fixed in Fedora 24 / Intel's drivers. It's still a bit fiddly/tricky to get that setup right, and with a wrong setup (failed autotuning) the performance is bad.

**Comment by bhack** [comment]
If you plan to integrate it in libdnn we could concentrate to libdnn integration. The problems is that are not clear what are the responsibilities that you want to handle in libdnn. If we need to handle other kernels ops for other layers outside libdnm we need all the Opencl/Cuda bootstrap pipeline, kernel source code retrieval,  program manager/program caching etc (that it is partially implemented also in Intel conv spatial with unique keys). I understand that you need to handle bootstrap, build and launch for kernel tuning/timing but we need to find a way to coordinate without managing to much replicated source code in multiple repositories.

**Comment by naibaf7** [comment]
@bhack 
It has been stated before: LibDNN is a collection for fast convolution implementations. Maybe I should update that in the repository description & readme.

You will need, as you noticed, the whole OpenCL/CUDA through CLCudaAPI, ViennaCL or similar, mainly because leaving this responsibility to the framework rather than the convolution engine makes sense, that's how Caffe, TensorFlow, Theano and so on are doing it. The reason is, among flexibility to use other libraries with the same abstraction level (which are cuDNN, cuBLAS, clBLAS and cuRAND, to name a few), also the memory management, which needs to happen at the tensor/blob levels of the framework.

You will also need an implementation of all kernels besides convolution, which concerns mostly pooling, activation and loss functions.
The reason these are NOT in LibDNN is that there is no autotuning going on in these kernels, so they can simply be added as single-source kernels for all devices. However, frameworks typically have different parameters and interfaces for these layers, so these are also left as a framework responsibility. Especially how the error / loss is handled may differ greatly. This issue was actually showcased in Caffe a few times, and it was recommended to use Caffe's own pooling implementation rathern than cuDNN's, see (https://github.com/BVLC/caffe/issues/2015, https://github.com/BVLC/caffe/issues/2688).

LibDNN itself therefore handles kernel generation, caching, loading and launching (for convolutions), and therefore exposes an interface similar to cuDNN to the deep learning frameworks. The expected inputs are valid OpenCL or CUDA memory objects and the convolution parameters.

It makes sense to include Intel's spatial kernels because this should obviously be usable outside of Caffe, and the autotuning could take care of selecting the fastest kernel within LibDNN. This has been an issue we noticed in Caffe, since people were confused on how to find the fastest implementation for their devices, with many compile options and runtime flags available.

As to time of implementation: Intel's kernel will move into the LibDNN abstraction as soon as the driver/beignet issues are fixed and the kernels are actually working on my development platform. I don't know yet when this will be the case though, but I am hoping it is within the next 2 months.

I hope this is clear enough explanation :)

**Comment by bhack** [comment]
OK I think we need to duplicate many libdnn functionalities if we want to have alternative implementations for other kernels/layers (other than convolution). But seems that we cannot get a better solution.

**Comment by bhack** [comment]
And probably we need to wait next Fedora release to have a stable stack of Beignet if you are developing on that platform.

**Comment by naibaf7** [comment]
@bhack 
It should be fixed in the next few kernel x11/xorg-intel updates for Fedora 24 I hope :)
otherwise if beignet fixes LLVM/Clang-3.8 compability I can also make it work by compiling manually.

But also for the end-user experience it would be much better if this gets fixed sooner rather than later in the distribution packages.

I don't think you need to duplicate much of the libdnn functionality in order to make it work. CLCudaAPI does most of it for you. You might need to duplicate a part of the Caffe functionality, but you'd probably want to adapt the OpenCL kernels a bit to exactly match your CPU operators in tinycnn.

Some design decisions to think about are if you want to store kernels externally as .cl files, or if you want to move them into C++ code or C++ string resources. I think for a header library you should probably leave them as .cl files and treat them the same way as header files when installing tinycnn, and load them as strings. But that's just my opinion.

**Comment by bhack** [comment]
I don't know what is the Fedora policy. But ubuntu generally don't push new upstream releases inside the same distro version. In some cases backport some patches if this solve some RC bug.

**Comment by naibaf7** [comment]
@bhack 
Fedora is most of the time on the cutting-edge of development, as it is the "test" plattform of RedHat before they move packages into RHEL. It already ships with the latest LLVM/Clang and GCC-6.1.1, long before Ubuntu will ever push these releases inside their distros (which is why I like and use Fedora).

**Comment by bhack** [comment]
I generally use a Debian testing/unstable in the development environment see https://packages.debian.org/search?keywords=beignet

Fedora doesn't seems so cutting edge. It is still on Beignet 1.1.1 https://apps.fedoraproject.org/packages/beignet.

Covering only convolutional kernels require that Tinycnn rebuild some kernel matching a caching mechanism. So it is the combo of tuning and the coverage of only Convolution that require third party to duplicate the stack cause I need many things that libdnn require for the other kernels (i.e. kernel retriving, program compiler options handling, caching etc..).

I think that you cannot compare to cudnn cause it is closed source. So many strategies to not full rely on cudnn it is to not let the frameworks to have a full closed source core without a possibility to influence the API design.

**Comment by naibaf7** [comment]
@bhack 
But you understand the issue of including all kernels from Caffe as well, right?
And it will add dependencies such as clBLAS and CLBlast directly to libdnn...
I'm just not sure if that's a good strategy for libDNN... @hughperkins also said we should limit it to convolutions & possibly other layers that require this sort of dynamic kernel configuration and autotuning, so yeah.
I mean in the end you could also link tinycnn with Caffe, if you really want "all kernels" + launch code, because that's what it would be then.

**Comment by bhack** [comment]
An alternative could be to only feed tuning interfaces, timing and kernel source codes for convolution from the library and let the third party to pilot this elements.

**Comment by bhack** [comment]
In this cases I think we will need to only spend some effort to standardize some kernel metadata (program compile options, vendor extension, etc.)
